hotcrp: init at 3.0b3

bollu opened the pull request:
## Description of changes


## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

bollu committed:
hotcrp: init at 3.0b3

Co-Authored By: itihas <sahiti93@gmail.com>

bollu commented:
I'm new to nixpkgs, and this package will probably have lots of stylistic issues. I will fix all of the formatting (thanks for the linting!).

Aleksanaa commented:
1. Please use nixpkgs style commit message (and also pr title): `hotcrp: init at 3.0b3` and `nixos/hotcrp: init`, and describe any other changes with that prefix.
2. Please squash your commits to the same package/module to one commit in one pr.
3. Please remove the prefix "v" from the version number.
4. Please add your contact information to [maintainer list](https://github.com/NixOS/nixpkgs/blob/master/maintainers/maintainer-list.nix) before using `maintainers` (this is required for the first pr)

bollu commented:
@Aleksanaa I have changed the PR title and the commit title. I am not sure what should be marked as `nixos/hotcrp: init`. I've added to the maintainer list as well.

Aleksanaa commented:
> @Aleksanaa I have changed the PR title and the commit title. I am not sure what should be marked as `nixos/hotcrp: init`. I've added to the maintainer list as well. 

I mean the **same package/module**, not all commits. You've added a nixos module and a package. "nixos/hotcrp" is the module you added.

Aleksanaa commented on pkgs/top-level/all-packages.nix:
```diff
@@ -26657,6 +26657,8 @@ with pkgs;
 
   gotosocial = callPackage ../servers/gotosocial { };
 
+  hotcrp = callPackage ../servers/web-apps/hotcrp { };
```
New packages should be placed in `pkgs/by-name`, thus no index is needed.

Aleksanaa commented on nixos/modules/services/web-apps/hotcrp.nix:
```diff
@@ -0,0 +1,199 @@
+    # 442: HTTPS
+    # 25, 465, 587: sendmail.
+    # 23: telnet
+    networking.firewall.allowedTCPPorts = [ 80 443 25 465 587 23 ];
```
You can't make your module occupy these ports by default.

Aleksanaa commented on pkgs/servers/web-apps/hotcrp/default.nix:
```diff
@@ -0,0 +1,29 @@
+{ stdenv, lib, fetchFromGitHub, php }:
```
Where's php being used?

Aleksanaa commented on pkgs/servers/web-apps/hotcrp/default.nix:
```diff
@@ -0,0 +1,29 @@
+    hash = "sha256-t3/auUMkMEY0FS2R5xu7NnaC5q2ZHRrpSHnBC7274as=";
+  };
+
+  installPhase = ''
```
Is this all needed to form a package? Where's `runHook postInstall`

Aleksanaa commented on nixos/modules/services/web-apps/hotcrp.nix:
```diff
@@ -0,0 +1,199 @@
+        cd ${dataDir} # everything below is in dataDir
+        ./lib/createdb.sh --batch ${key} --replace || (mv ${dataDir} ${dataDir}.old && exit 1)
+        chown -R ${cfg.user}:${cfg.group} . || (mv ${dataDir} ${dataDir}.old && exit 1)
+        chmod 777 -R  . || (mv ${dataDir} ${dataDir}.old && exit 1)
```
Why to use chmod 777?

Aleksanaa commented on nixos/modules/services/web-apps/hotcrp.nix:
```diff
@@ -0,0 +1,199 @@
+      script = ''
+        set -o xtrace
+        set -euo pipefail
+        whoami
```
Why here's a `whoami`?

Aleksanaa commented on nixos/modules/services/web-apps/hotcrp.nix:
```diff
@@ -0,0 +1,199 @@
+    in
+    nameValuePair (key + "-init") ({
+      wantedBy = [ "multi-user.target" ];
+      after = [ "phpfpm-${key}.service" "mysql.service" ]; # https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/web-servers/phpfpm/default.nix
```
`network.target`?

Aleksanaa commented on nixos/modules/services/web-apps/hotcrp.nix:
```diff
@@ -0,0 +1,199 @@
+    nameValuePair (key + "-init") ({
+      wantedBy = [ "multi-user.target" ];
+      after = [ "phpfpm-${key}.service" "mysql.service" ]; # https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/web-servers/phpfpm/default.nix
+      serviceConfig = { User = "root"; Type = "oneshot"; }; # TODO: change to `cfg.user` and remove the `chown` below.
```
Why to explicitly use `root` user?

Aleksanaa commented on nixos/modules/services/web-apps/hotcrp.nix:
```diff
@@ -0,0 +1,199 @@
+    enable = mkEnableOption (lib.mdDoc "Hotcrp web application");
+    package = lib.mkPackageOption pkgs "hotcrp" { };
+    user = mkOption {
+      type = lib.types.str;
```
You've written `with lib;` above.

Aleksanaa commented on nixos/modules/services/web-apps/hotcrp.nix:
```diff
@@ -0,0 +1,199 @@
+                nameValuePair (key + ".*") "ALL PRIVILEGES"
+                ) cfg.instances;
+              } ];
+            };
```
Could you make your left and right brackets more symmetrical?

nu-nu-ko commented on nixos/modules/services/web-apps/hotcrp.nix:
```diff
@@ -0,0 +1,199 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
```
```suggestion
```
to address #208242 for this module, uses of `with` such as this should be avoided.
consider `inherits` where you find yourself using `lib.foo` excessively.
[this](https://nix.dev/guides/best-practices.html#with-scopes)  nix.dev page on best practices with `with` may be helpful

nu-nu-ko commented on nixos/modules/services/web-apps/hotcrp.nix:
```diff
@@ -0,0 +1,199 @@
+  {
+  # Interface
+  options.services.hotcrp = {
+    enable = mkEnableOption (lib.mdDoc "Hotcrp web application");
```
```suggestion
    enable = mkEnableOption "Hotcrp web application";
```
`lib.mdDoc` is now just an alias and can be safely removed everywhere.
see https://github.com/NixOS/nixpkgs/commit/d36f950d405369abca5144b14a201bc25c4a080a and https://github.com/NixOS/nixpkgs/pull/237557
